org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/persistence/DocumentProvider.xtend

```diff
@@ -137,26 +174,60 @@ class DocumentProvider {
 		}
 
 	}
+	
+	private def void repoSync(Consumer<StageState> handler) {
+		val mergeConflictState = pull
+
+		if (!mergeConflictState.isPresent) {
+			push
+		} else {
+			handler.accept(mergeConflictState.get)
+		}
+	}
+
+	private def void resetToRemoteAndCreateBackup(StageState mergeConflictState, String resourcePath, String content) {
+		resetToRemoteState
+		var exceptionMessage = mergeConflictState.getConflictMessage(resourcePath)
+		if (!content.isNullOrEmpty) {
+			val backupFileName = createLocalBackup(resourcePath, content)
+			
+			exceptionMessage = exceptionMessage.appendBackupNote(backupFileName)
+		}
+		throw new ConflictingModificationsException(exceptionMessage)
+	}
+	
+	private def createLocalBackup(String resourcePath, String content) {
+		val workspace = workspaceProvider.workspace
+		var fileSuffix = BACKUP_FILE_SUFFIX
+		var backupFile = new File(workspace, resourcePath + fileSuffix)
+		var numberSuffix = 0
```

Comment:
How about something like
val numberSuffix = (0..9).findFirst[ i | !new File(...).exists ]

Looks nicer, is bounded to ten tries ...